# (실습) 넘파이 어레이

**주의사항**

- `None`은 적절한 표현식으로 대체한다.
- `Pass`는 적절한 명령문으로 대체한다.

## numpy 모듈

**문제 1**

아래 코드를 설명하라.

In [107]:
import numpy as np

`코드설명`: numpy 모듈을 np 로 불러옴.

아래 코드는 수정하지 마세요!!!

In [108]:
np.__version__

'1.26.4'

**문제 2**

아래 코드의 기능을 설명하라

In [109]:
np.random.seed(42)

`코드설명`: 난수를 생성할 때 사용되는 시드를 42로 정함.



**문제 3**

부동소수점으로 구성된 어레이의 항목은 기본적으로 소수점 이하 8자리까지 출력한다.

In [110]:
Z = np.random.random((40,40))
Z

array([[0.37454012, 0.95071431, 0.73199394, ..., 0.09767211, 0.68423303,
        0.44015249],
       [0.12203823, 0.49517691, 0.03438852, ..., 0.07404465, 0.35846573,
        0.11586906],
       [0.86310343, 0.62329813, 0.33089802, ..., 0.18657006, 0.892559  ,
        0.53934224],
       ...,
       [0.81783422, 0.12020906, 0.5444891 , ..., 0.89803055, 0.2434823 ,
        0.92703454],
       [0.06026739, 0.93443603, 0.35162269, ..., 0.41102847, 0.8398613 ,
        0.90002312],
       [0.35342138, 0.23687056, 0.78052552, ..., 0.07941937, 0.08565825,
        0.89419087]])

`np.set_printoptions()` 의 기능을 다음 두 개의 예를 이용하여 설명하라.

In [111]:
np.set_printoptions(precision=4)

In [112]:
Z

array([[0.3745, 0.9507, 0.732 , ..., 0.0977, 0.6842, 0.4402],
       [0.122 , 0.4952, 0.0344, ..., 0.074 , 0.3585, 0.1159],
       [0.8631, 0.6233, 0.3309, ..., 0.1866, 0.8926, 0.5393],
       ...,
       [0.8178, 0.1202, 0.5445, ..., 0.898 , 0.2435, 0.927 ],
       [0.0603, 0.9344, 0.3516, ..., 0.411 , 0.8399, 0.9   ],
       [0.3534, 0.2369, 0.7805, ..., 0.0794, 0.0857, 0.8942]])

In [113]:
np.set_printoptions(precision=6)

In [114]:
Z

array([[0.37454 , 0.950714, 0.731994, ..., 0.097672, 0.684233, 0.440152],
       [0.122038, 0.495177, 0.034389, ..., 0.074045, 0.358466, 0.115869],
       [0.863103, 0.623298, 0.330898, ..., 0.18657 , 0.892559, 0.539342],
       ...,
       [0.817834, 0.120209, 0.544489, ..., 0.898031, 0.243482, 0.927035],
       [0.060267, 0.934436, 0.351623, ..., 0.411028, 0.839861, 0.900023],
       [0.353421, 0.236871, 0.780526, ..., 0.079419, 0.085658, 0.894191]])

`코드설명`: 소수점 자릿수를 제한함함. 지정된 자릿수의 위치에서 반올림함.

아래 코드는 기본 옵션값으로 되돌린다. 수정하지 마세요.

In [115]:
np.set_printoptions(precision=8)

## 1차원 어레이

**문제 1**

크기가 10이고 모든 항목이 0인 1차원 어레이를
변수 `Z`가 가리키도록 하라.
단, `np.zeros()` 함수를 반드시 이용한다.

In [116]:
# None을 적절한 표현식으로 대체하라.
Z = np.zeros(10)

아래 코드는 수정하지 마세요!!!

In [117]:
assert (Z == np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])).all()

**문제 2**

변수 `Z`가 10에서 49까지의 정수 중에서 짝수로만 구성된 1차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수를 반드시 이용한다.

In [118]:
# None을 적절한 표현식으로 대체하라.
Z =  np.arange(10,50, 2)
Z

array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42,
       44, 46, 48])

아래 코드는 수정하지 마세요!!!

In [119]:
assert (Z == np.array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42,
                        44, 46, 48])).all()

**문제 3**

변수 `Z`가 아래와 같이 선언되었다.

In [120]:
Z = np.array([2, 3, 1, 0, 4])
Z

array([2, 3, 1, 0, 4])

변수 `Z`의 항목을 정렬하라.
단, sort() 메서드를 활용한다.

In [121]:
# None을 적절한 표현식으로 대체하라.
Z.sort()
Z

array([0, 1, 2, 3, 4])

아래 코드는 수정하지 마세요!!!

In [122]:
assert (Z == np.arange(5)).all()

**문제 4**

변수 `Z`가 크기 30이며 무작위로 생성된 1차원 어레이를 가리키도록 하라.
또한 포함된 항목들의 평균값을 계산하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
`np.random.random()` 함수, `mean()` 어레이 메서드를 반드시 이용한다.

In [123]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random(30)
Z_mean = Z.mean()

아래 코드는 수정하지 마세요!!!

In [124]:
assert (Z.shape == (30,))
print(f"1차원 어레이 평균값: {Z_mean}")

1차원 어레이 평균값: 0.4618868095723934


**문제 5**

1차원 어레이가 다음과 같이 주어졌다.

In [125]:
Z = np.arange(1, 4)

(1) 아래 표현식의 계산 결과를 설명하라.

In [126]:
Z*Z

array([1, 4, 9])

`코드설명`: 항목별 곱셈셈

(2) 아래 표현식의 계산 결과를 설명하라.

In [127]:
Z**Z

array([ 1,  4, 27])

`코드설명`: 항목별 지수 승

(3) 아래 표현식의 계산 결과를 설명하라.

In [128]:
Z < -Z

array([False, False, False])

`코드설명`: 항목별 크기 비교교

(4) 아래 표현식의 계산 결과를 설명하라.

In [129]:
Z/1/1

array([1., 2., 3.])

`코드설명`: 항목별 나눗셈 (2회 실행함.)

**문제 6**

변수 `Z`가 아래 모양의 어레이를 가리키도록 하라.
단, `np.linspace()` 함수와 `endpoint` 키워드 인자를 이용한다.

```
[0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
```

In [130]:
# None을 적절한 표현식으로 대체하라.
Z = np.linspace(0, 1, 10, endpoint=False)
Z

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])

아래 코드는 수정하지 마세요!!!

In [131]:
assert np.isclose(Z, np.array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])).all()

**문제 7**

아래 코드의 실행결과를 설명하라.

In [132]:
Z = np.zeros((10,10))
print("%d bytes" % (Z.size * Z.itemsize))

800 bytes


`코드설명`: Z.dtype= float64, 8 byte의 메모리가 필요함. 
Z에 100개의 항목이 있기때문에 총 800 byte의 메모리가 어레이 저장에 사용됨됨.

**문제 8**

다음 리스트에서 0이 아닌 항목들의 인덱스로 구성된 어레이를 한줄 코드로 생성하라. 
단. `np.nonzero()` 함수를 활용한다.

In [133]:
Z = [1, 2, 0, 0, 4, 0]

In [134]:
# None을 적절한 표현식으로 대체하라.

non_zeros = np.nonzero(Z)

아래 코드는 수정하지 마세요!!!

In [135]:
assert (non_zeros == np.array([0, 1, 4])).all()

**문제 9**

두 개의 1차원 어레이가 다음과 같이 주어졌다.

In [136]:
A = np.ones(3)*1
B = np.ones(3)*2

아래 코드의 실행 결과를 설명하라.

In [137]:
(A+B)*(-A/2)

array([-1.5, -1.5, -1.5])

`코드설명`: 항목간 곱셈과 나눗셈을 실행한 것을 곱함.

In [138]:
np.add(A,B,out=B)

array([3., 3., 3.])

`코드설명`: B에 A와 B가 더해진 값이 할당됨.

In [139]:
np.divide(A,2,out=A)

array([0.5, 0.5, 0.5])

`코드설명`: A에 A 나누기 2가 할당됨.

In [140]:
np.negative(A,out=A)

array([-0.5, -0.5, -0.5])

`코드설명`: A에 A를 부정한 결과가 할당됨.

In [141]:
np.multiply(A,B,out=A)

array([-1.5, -1.5, -1.5])

`코드설명`: A에 A와 B의 항목별 곱셈결과가 할당됨.

## 2차원 어레이

**문제 1**

변수 `Z`가 (5, 5) 모양의 2차원 어레이를 가리키도록 하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
또한 `np.random.random()` 함수를 이용한다.

In [142]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random((5,5))

아래 코드는 수정하지 마세요!!!

In [143]:
assert (Z.shape == (5, 5))
assert (Z.dtype == np.float64)

**문제 2**

변수 `Z`가 0에서 8까지의 정수로 이루어진 (3, 3) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수와 `reshape()` 어레이 메서드를 반드시 이용한다.

In [144]:
# None을 적절한 표현식으로 대체하라.
Z = np.arange(9).reshape(3, 3)

아래 코드는 수정하지 마세요!!!

In [145]:
assert (Z == np.array([[0, 1, 2],
                       [3, 4, 5],
                       [6, 7, 8]])).all()

**문제 3**

변수 `Z`가 (3, 3) 모양의 2차원 어레이를 가리키도록 하라.
단, 2차원 어레이의 대각선 상은 1이고 나머지는 0이어야 한다.
또한 `np.eye()` 함수를 반드시 이용한다.

In [146]:
Z = np.eye(3)

아래 코드는 수정하지 마세요!!!

In [147]:
assert (Z == np.array([[1., 0., 0.],
                       [0., 1., 0.],
                       [0., 0., 1.]])).all

**문제 4**

변수 `Z`가 아래와 같은 (5,5) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수와 `np.diag()` 함수의 `k` 키워드 인자를 반드시 활용한다.

```
[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]
```

In [148]:
# None을 적절한 표현식으로 대체하라.
Z = np.diag(1+np.arange(4),k=-1) 

아래 코드는 수정하지 마세요!!!

In [149]:
assert (Z == np.array([[0, 0, 0, 0, 0],
                        [1, 0, 0, 0, 0],
                        [0, 2, 0, 0, 0],
                        [0, 0, 3, 0, 0],
                        [0, 0, 0, 4, 0]])).all()

**문제 5**

변수 `Z`가 아래와 같은 (5, 5) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.zeros()` 함수, `np.arange()` 함수, 2차원 어레이 덧셈을 이용한다.

```
[[0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]]
```

In [150]:
# None을 적절한 표현식으로 대체하라.
Z = np.zeros((5,5)) + np.arange(5)

아래 코드는 수정하지 마세요!!!

In [151]:
assert (Z == np.array([[0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.]])).all()

**문제 6**

아래 3x3 단위 행렬에 해당하는 2차원 어레이를 한줄 코드로 생성하라.
단, `np.eye()` 함수를 이용한다.

```
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
```

In [152]:
# None을 적절한 표현식으로 대체하라.
Z = np.eye(3)

아래 코드는 수정하지 마세요!!!

In [153]:
assert (Z == np.array([[1., 0., 0.],
                       [0., 1., 0.],
                       [0., 0., 1.]])).all()

**문제 7**

무작위 수로 구성된 (5, 5) 모양의 2차원 어레이가 다음과 같다.

In [154]:
Z = np.random.random((5, 5))

변수 `Z`가 가리키는 2차원 어레이에 대해 min-max 크기조정(정규화)을 적용하라.

min-max 크기조정(정규화)은 항목의 모든 값을 아래 식을 통해 0과 1 사이의 부동소수점으로 변환한다. $max$와 $min$은 각각 최대 항목과 최소 항목을 가리킨다.

$$
\frac{x-min}{max - min}
$$

In [155]:
Z = (Z - Z.min()) / (Z.max() - Z.min())
Z

array([[0.93942291, 0.51894519, 0.24648319, 0.34684899, 0.40196191],
       [0.57029517, 0.47710894, 0.92014602, 0.45656993, 1.        ],
       [0.84555852, 0.14001954, 0.93244054, 0.03889992, 0.80827384],
       [0.32889736, 0.87469398, 0.86381031, 0.79562698, 0.77818373],
       [0.24555569, 0.        , 0.35343197, 0.10162565, 0.67384296]])

아래 코드는 수정하지 마세요!!!

In [156]:
assert (Z.max() == 1 and Z.min() == 0)

## 3차원 어레이

**문제 1**

(1) 변수 `Z`가 (3, 3, 3) 모양의 3차원 어레이를 가리키도록 하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
또한 `np.random.random()` 함수를 이용한다.

In [157]:
Z = np.random.random((3,3,3))

아래 코드는 수정하지 마세요!!!

In [158]:
assert (Z.shape == (3, 3, 3))
assert (Z.dtype == 'f8')

(2) `Z`에 포함된 항목들 중에서 최댓값과 최솟값의 차이를 구하라.
단, `maximum`, `mininum` 두 변수가 각각 최댓값과 최솟값을 가리키도록 하며,
`max()`와 `min()` 어레이 메서드를 이용한다.

In [159]:
maximum = Z.max()
minimum = Z.min()

아래 코드는 수정하지 마세요!!!

In [160]:
print(f"최댓값과 최솟값의 차이: {maximum - minimum}")

최댓값과 최솟값의 차이: 0.9497982939104279


## `dtype` 종류 

**문제 1**

아래 두 코드의 실행결과를 설명하라.

In [161]:
for dtype in [np.int8, np.int32, np.int64]:
    print(np.iinfo(dtype).min)
    print(np.iinfo(dtype).max)

-128
127
-2147483648
2147483647
-9223372036854775808
9223372036854775807


`코드설명`: 정수형 자료에 허용되는 비트수 최솟값과 최댓값
8비트: -128~127 정수
16비트: : -2147483648~2147483647 정수
32비트: -9223372036854775808~9223372036854775807 정수

In [162]:
for dtype in [np.float32, np.float64]:
    print(np.finfo(dtype).min)
    print(np.finfo(dtype).max)
    print(np.finfo(dtype).eps)

-3.4028235e+38
3.4028235e+38
1.1920929e-07
-1.7976931348623157e+308
1.7976931348623157e+308
2.220446049250313e-16


`코드설명`: 부동소수점 자료에 허용된 비트수 최솟/최댓값

**문제 2**

부동소수점으로 구성된 1차원 어레이가 아래와 같이 주어졌다.

In [163]:
Z = (np.random.rand(10)*100).astype(np.float32)

변수 `Y`가 변수 `Z`가 가리키는 1차원 어레이의 dtype이 `int32`로 형변환된 1차원 어레이를 가리키도록 하라.
단, `astype()` 메서드를 이용한다.

In [166]:
# None을 적절한 표현식으로 대체하라.
Y = Z.astype(np.int32)
Y

array([21, 61, 65, 15,  6, 78, 45,  5, 99,  5])

아래 코드는 수정하지 마세요!!!

In [167]:
assert (Y.shape == (10,))
assert (Y.dtype == np.int32)

**문제 3**

변수 `Z`가 아래와 같이 무작위로 생성된 길이가 10인 1차원 어레이를 가리킨다.

In [ ]:
Z = np.random.uniform(0,10,10)
Z

(1) 나머지 연산자(`%`)를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라. 

In [168]:
Y = Z - Z%1
Y

array([21., 61., 65., 15.,  6., 78., 45.,  5., 99.,  5.], dtype=float32)

아래 코드는 수정하지 마세요!!!

In [169]:
assert (Y == Z.astype(np.int32)).all()

(2) 몫 연산자(`//`)를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [170]:
Y = Z // 1

아래 코드는 수정하지 마세요!!!

In [171]:
assert (Y == Z.astype(np.int32)).all()

(3) 내림 함수 `np.floor()` 함수를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라. 

In [172]:
Y = np.floor(Z)

아래 코드는 수정하지 마세요!!!

In [173]:
assert (Y == Z.astype(np.int32)).all()

(4) `np.trunc()` 함수를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라. 

In [174]:
Y = np.trunc(Z)

아래 코드는 수정하지 마세요!!!

In [175]:
assert (Y == Z.astype(np.int32)).all()